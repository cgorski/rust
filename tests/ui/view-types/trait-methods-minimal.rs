#![feature(view_types)]
#![allow(incomplete_features)]

// RESTRICTION TEST: Trait methods with view types are not supported
// This test verifies that both trait definitions and trait impls with view types
// are properly rejected with clear error messages (not ICE)

// Test 1: Simplest possible trait with view type
trait Counter {
    fn increment(&{mut count} mut self); //~ ERROR patterns aren't allowed in functions without bodies
    //~| WARN this was previously accepted by the compiler but is being phased out
}

struct MyCounter {
    count: usize,
    name: String,
}

impl Counter for MyCounter {
    fn increment(&{mut count} mut self) { //~ ERROR view types are not supported on trait method implementations
        self.count += 1;
    }
}

fn main() {
    let mut counter = MyCounter {
        count: 0,
        name: String::from("test"),
    };

    // Call trait method directly
    counter.increment();

    println!("Count: {}", counter.count);
}
