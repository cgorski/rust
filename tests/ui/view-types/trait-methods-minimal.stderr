error: patterns aren't allowed in functions without bodies
  --> $DIR/trait-methods-minimal.rs:10:18
   |
LL |     fn increment(&{mut count} mut self);
   |                  ^^^^^^^^^^^^^^^^^^^^^ help: remove `mut` from the parameter: `self`
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #35203 <https://github.com/rust-lang/rust/issues/35203>
   = note: `#[deny(patterns_in_fns_without_body)]` (part of `#[deny(future_incompatible)]`) on by default

error: view types are not supported on trait method implementations
  --> $DIR/trait-methods-minimal.rs:20:18
   |
LL |     fn increment(&{mut count} mut self) {
   |                  ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: view types are restricted to inherent methods
   = note: trait method support requires trait system integration
   = help: consider using regular references without view type annotations

error: aborting due to 2 previous errors

